src/plugins/default/markbind-plugin-plantuml.js

```diff
@@ -22,74 +23,80 @@ const processedDiagrams = new Set();
 
 /**
  * Generates diagram and returns the file name of the diagram
- * @param fileName name of the file to be generated
+ * @param imageOutputPath output path of the diagram to be generated
  * @param content puml dsl used to generate the puml diagram
- * @param config sourcePath and resultPath from parser context
- * @returns {string} file name of diagram
  */
-function generateDiagram(fileName, content, config) {
-  const { resultPath } = config;
-
-  const outputDir = path.join(path.dirname(resultPath), path.dirname(fileName));
-  // Path of the .puml file
-  const outputFilePath = path.join(outputDir, path.basename(fileName));
-  // Tracks built files to avoid accessing twice
-  if (processedDiagrams.has(outputFilePath)) { return fileName; }
-  processedDiagrams.add(outputFilePath);
+function generateDiagram(imageOutputPath, content) {
+  // Avoid generating twice
+  if (processedDiagrams.has(imageOutputPath)) return;
```

Comment:
> I'm a little hesitant about using the event loop and setImmediate to defer the puml image generation to the end of the build process. Depending on how other callbacks are handled in the poll queue, the behaviour might not be as we expect (please correct me if I'm wrong here, it's been a while since I had to look at the node event loop!).

I think we should be good here 😬, https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/, `setImmediate` is executed in the `check` phase, only after all callbacks (ie. all of our `page.generate()`s) have resolved.

Moving part of it to `postRender` works a well, but to a lesser extent as you mentioned. It also involves some extra parsing work as we can only move part and not the whole plugin to `postRender`

Reply:
The ideal solution would to be find some way to let prioritise the resource cleanup callbacks (im guessing that's how node does it) after the puml processes have finished. There dosen't seem to be an interface for this after hours of searching though, which makes sense since node's event loop wasn't really built for continuous execution of a massive number of callbacks.

An alternative solution, is that if and when we manage to parallelise page generation, we could change the execution back to synchronous. This way we have a good amount of thread usage still 🎉